webrtc/stun/stun_connection.nim

```diff
@@ -0,0 +1,249 @@
+# Nim-WebRTC
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import strutils
+import bearssl, chronos, chronicles, stew/[objects, byteutils]
+import ../[udp_connection, errors], stun_message, stun_attributes
+
+logScope:
+  topics = "webrtc stun stun_connection"
+
+# TODO:
+# - Need to implement ICE-CONTROLL(ED|ING) for browser to browser (not critical)
+
+const
+  StunBindingRequest* = 0x0001'u16
+  StunBindingResponse* = 0x0101'u16
+  StunBindingErrorResponse* = 0x0111'u16
+
+type
+  StunUsernameProvider* = proc(): string {.raises: [], gcsafe.}
+  StunUsernameChecker* = proc(username: seq[byte]): bool {.raises: [], gcsafe.}
+  StunPasswordProvider* = proc(username: seq[byte]): seq[byte] {.raises: [], gcsafe.}
+
+  StunConn* = ref object
+    conn*: UdpConn # Underlying UDP connexion
+    laddr*: TransportAddress # Local address
+    raddr*: TransportAddress # Remote address
+    dataRecv*: AsyncQueue[seq[byte]] # data received which will be read by DTLS
+    stunMsgs*: AsyncQueue[seq[byte]] # stun messages received and to be
```

Comment:
is this queue unbounded?

Reply:
Yes, those two queues are unbounded.

Reply:
should they be?

Reply:
I'm not sure this is something always mentioned on specs, but I believe there should always be a limit to avoid memory DoS attacks. See more in https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#bound-everything.

Reply:
done